File: solutions/kotlin/largest-series-product/1/src/main/kotlin/Series.kt
class Series(val input : String){

    private val digits : List<Long>

    init {
        // validity check
        if(input.any { !Character.isDigit(it) } )
            throw IllegalArgumentException()

        // map all characters to Long
        digits = input.map { it.toString().toLong() }
    }

    private fun product( digits : List<Long>) = digits.fold(1L) { acc, i -> acc * i }

    private fun <T> List<T>.windowed(span: Int) = (0 .. (this.size - span)).map { this.subList(it, it + span) }

    fun getLargestProduct(n : Int) : Long{
        // validity check
        if(n < 0 || n > digits.size){
            throw IllegalArgumentException()
        }

        return digits.windowed(n).map { product(it) }.max() ?: 0
    }
}

File: solutions/go/series/1/series.go
package series

func All(n int, s string) []string {
	if n > len(s) {
		return nil
	}
	series := []string{}
	for i := 0; i < len(s)-n+1; i++ {
		series = append(series, s[i:i+n])
	}
	return series
}

func UnsafeFirst(n int, s string) string {
	return All(n, s)[0]
}

func First(n int, s string) (string, bool) {
	series := All(n, s)
	if len(series) == 0 {
		return "", false
	}
	return series[0], true
}
